File: bigseqkit/subseq.go
package bigseqkit

import "ignis/driver/api"

type SeqKitSubseqOptions struct {
	inner SubseqOptions
}

type SubseqOptions struct {
	Config     KitConfig
	Chr        *[]string
	Region     *string
	Gtf        *string
	Feature    *[]string
	UpStream   *int
	DownStream *int
	OnlyFlank  *bool
	Bed        *string
	GtfTag     *string
}

func (this *SubseqOptions) setDefaults() *SubseqOptions {
	this.Config.setDefaults()
	setDefault(&this.Chr, []string{})
	setDefault(&this.Region, "")
	setDefault(&this.Gtf, "")
	setDefault(&this.Feature, []string{})
	setDefault(&this.UpStream, 0)
	setDefault(&this.DownStream, 0)
	setDefault(&this.OnlyFlank, false)
	setDefault(&this.Bed, "")
	setDefault(&this.GtfTag, "")

	return this
}

func (this *SeqKitSubseqOptions) Config(v *SeqKitConfig) *SeqKitSubseqOptions {
	this.inner.Config = v.inner
	return this
}

func (this *SeqKitSubseqOptions) Chr(v []string) *SeqKitSubseqOptions {
	this.inner.Chr = &v
	return this
}

func (this *SeqKitSubseqOptions) Region(v string) *SeqKitSubseqOptions {
	this.inner.Region = &v
	return this
}

func (this *SeqKitSubseqOptions) Gtf(v string) *SeqKitSubseqOptions {
	this.inner.Gtf = &v
	return this
}

func (this *SeqKitSubseqOptions) Feature(v []string) *SeqKitSubseqOptions {
	this.inner.Feature = &v
	return this
}

func (this *SeqKitSubseqOptions) UpStream(v int) *SeqKitSubseqOptions {
	this.inner.UpStream = &v
	return this
}

func (this *SeqKitSubseqOptions) DownStream(v int) *SeqKitSubseqOptions {
	this.inner.DownStream = &v
	return this
}

func (this *SeqKitSubseqOptions) OnlyFlank(v bool) *SeqKitSubseqOptions {
	this.inner.OnlyFlank = &v
	return this
}

func (this *SeqKitSubseqOptions) Bed(v string) *SeqKitSubseqOptions {
	this.inner.Bed = &v
	return this
}

func (this *SeqKitSubseqOptions) GtfTag(v string) *SeqKitSubseqOptions {
	this.inner.GtfTag = &v
	return this
}

func Subseq(input *api.IDataFrame[string], o *SeqKitSubseqOptions) (*api.IDataFrame[string], error) {
	if o == nil {
		o = &SeqKitSubseqOptions{}
	}
	opts := o.inner
	opts.setDefaults()

	libprepare, err := api.AddParam(libSource("SubseqTransform"), "opts", OptionsToString(opts))
	if err != nil {
		return nil, err
	}

	return api.MapPartitions[string, string](input, libprepare)
}

File: bigseqkit-py/bigseqkit/subseq.py
from typing import List

from bigseqkit.helper import _setDefault, _libSource, _config, _optionsToString, _parseKargs, SeqKitConfig, IDataFrame


class SeqKitSubseqOptions:

    def __init__(self):
        self.__inner = SubseqOptions()

    def config(self, v: SeqKitConfig):
        self.__inner.Config = _config(v)

    def chr(self, v: List[str]):
        self.__inner.Chr = v

    def region(self, v: bool):
        self.__inner.Region = v

    def gtf(self, v: bool):
        self.__inner.Gtf = v

    def feature(self, v: List[str]):
        self.__inner.Feature = v

    def upStream(self, v: int):
        self.__inner.UpStream = v

    def onlyFlank(self, v: bool):
        self.__inner.OnlyFlankOnlyFlank = v

    def bed(self, v: str):
        self.__inner.Bed = v

    def gtfTag(self, v: str):
        self.__inner.GtfTag = v

    def _run(self, input: IDataFrame, **kwargs):
        opts = self.__inner
        _parseKargs(opts, kwargs)
        opts.setDefaults()
        libprepare = _libSource("SubseqTransform").addParam("opts", _optionsToString(opts))
        return input.mapPartitions(libprepare)


class SubseqOptions:

    def __init__(self):
        self.Config = None  # KitConfig
        self.Chr = None  # list[str]
        self.Region = None  # str
        self.Gtf = None  # str
        self.Feature = None  # list[str]
        self.UpStream = None  # int
        self.DownStream = None  # int
        self.OnlyFlank = None  # bool
        self.Bed = None  # str
        self.GtfTag = None  # str

    def setDefaults(self):
        _setDefault(self, "Config", _config(SeqKitConfig())).setDefaults()
        _setDefault(self, "Chr", [])
        _setDefault(self, "Region", "")
        _setDefault(self, "Gtf", "")
        _setDefault(self, "Feature", [])
        _setDefault(self, "UpStream", 0)
        _setDefault(self, "DownStream", 0)
        _setDefault(self, "OnlyFlank", False)
        _setDefault(self, "Bed", "")
        _setDefault(self, "GtfTag", "")


def subSeq(input: IDataFrame, o: SeqKitSubseqOptions = None, **kwargs):
    if o is None:
        o = SeqKitSubseqOptions()
    return o._run(input, **kwargs)
